### PR TITLE
[storage/journal] Durably record the pruning boundary for recovery

### PR DESCRIPTION
**Pull request by roberto-bayardo**
## Summary

Pruning derives its boundary from whichever old blobs happen to survive on disk, so recovery cannot distinguish a legitimately pruned prefix from lost retained data: an interrupted prune — or an externally corrupted survivor below the boundary — is silently reclassified as pruned history, and the journal reopens as a shorter but seemingly valid state. This PR records the pruning boundary durably **before** removing any blob, so recovery *knows* the boundary instead of inferring it.

Rebased onto and built on top of #4310 (durability `Barrier` + best-effort recovery watermark in `start_sync`); it keeps #4310's prune structure and consuming API and adds the durable boundary + recovery hardening on top, rather than replacing them.

## What changed

- **`persist` records every boundary.** `Checkpoint::persist` now records all pruning boundaries, including blob-aligned ones, dropping the aligned-drop optimization that inferred aligned boundaries from the oldest surviving blob. Inference is what makes lost retained data indistinguishable from a legitimate prune; recording the boundary removes that ambiguity.
- **Prune commits the boundary before removal, with a watermark that covers it.** The fixed journal persists the boundary to its checkpoint before removing blobs; the variable journal prunes the offsets journal (its durable prune record) before removing data. The prune first syncs all data durable, then records that proven size (`bounds.end`) as the recovery watermark — so the just-synced retained items are covered rather than left below a stale watermark. A committed boundary lets recovery complete an interrupted prune, so blob removal no longer has to finish atomically or in order. This composes with #4310's `Barrier`: the pre-removal data sync marks the barrier durable, and the boundary is committed after it.
- **Recovery treats the recorded boundary as authoritative.** A boundary whose blob is missing while later blobs survive is reported as corruption — uniformly for aligned and mid-blob boundaries — instead of being masked as pruned history or silently adopting the oldest survivor. The verdict is reached before any surviving blob is removed and is stable across retried inits. Recovery also truncates a torn interior page in the two newest blobs (the only ones that can hold non-durable data), and completes interrupted prunes and clears.

## Recovery semantics

With a recorded boundary, init reconciles physical state against it:

- blobs below the boundary are removed, completing an interrupted prune;
- the boundary is recorded before any removal and removals never touch the boundary blob, so a missing boundary blob with surviving later blobs is **always** corruption — for both aligned and mid-blob boundaries — a loss state that inference previously masked as pruned history (and that the mid-blob path previously adopted silently);
- the prune records a watermark covering the data it just made durable, so losing that retained data after a prune is corruption rather than a journal that reopens empty (a stale watermark previously under-recorded durability here);
- variable journals use the offsets journal's pruning boundary as the committed data-prune record; because the offsets boundary is committed before data removal, a crash mid-removal leaves the data prefix ahead of the boundary, which recovery finishes removing (offsets-ahead-of-data is an interrupted prune, not corruption);
- data ahead of the committed variable-journal boundary — including a missing boundary blob — is corruption, as is a recovery watermark beyond the recoverable size.

Corruption verdicts leave surviving blobs untouched and are stable across retried inits **for crash-reachable states**: the pre-reconcile trailing-byte trim runs before the boundary is validated, so a synced (chunk-aligned) blob is never trimmed by an in-model crash; a blob it mutates despite a later corruption verdict implies external corruption of already-durable data, outside the crash model.

## Compatibility

A recorded pruning boundary is authoritative, which changes one upgrade case: a journal that crashed **mid-prune under a pre-#4235 binary** can leave a recorded boundary that lags physical storage (an `init_at_size` mid-blob boundary whose blob the old prune removed before the next sync persisted the update). Old code adopted the surviving blobs; this PR reports corruption. **Sync/drain cleanly before upgrading from a pre-#4235 binary.** Journals with no recorded boundary at all still derive it from the oldest blob on first init and then persist it explicitly, so aligned upgrades (where #4310 recorded no boundary) are unaffected.

## Costs

- Each prune performs one durable metadata sync (fixed) / offsets prune (variable) before removing blobs; previously an aligned prune wrote no boundary metadata and left its boundary implicit in which blobs survived.
- The first init after upgrade persists the derived boundary once.

## Validation

- `just test -p commonware-storage` — full suite (2865 tests), including failing-first prune-watermark regressions (`test_fixed_recovery_prune_watermark_covers_retained_data`, `test_variable_recovery_prune_watermark_covers_retained_data`) and the uniform missing-boundary corruption test (`test_fixed_recovery_rejects_lagging_mid_blob_boundary`, a real `init_at_size` journal with its boundary blob removed).
- `just test-conformance -p commonware-storage` — 128 tests pass. The always-recorded-boundary change regenerated the contiguous and journal-backed qmdb/adb fixtures earlier in this PR; the recovery fixes moved no further hashes (the watermark is internal recovery state).
- `just fix-fmt`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01WSarn1C4iMrjHQHrVcanFo

### CODE REVIEW

**Comment by roberto-bayardo** [comment]
bugbot run

File: storage/src/journal/contiguous/fixed.rs [hunk 1]
```diff
@@ -1053,6 +1052,13 @@ impl<E: Context, A: CodecFixedShared> Journal<E, A> {
         }
 
         let new_boundary = super::blob_first_position(min_blob, self.items_per_blob.get())?;
+        // Commit the boundary before removing blobs. Recovery treats an ahead boundary as an
+        // interrupted prune and completes it, so blob removal never has to finish atomically
+        // or in order.
+        let recovery_watermark = self.recovery_watermark();
+        self.checkpoint
+            .persist(new_boundary, recovery_watermark)
+            .await?;
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs:1142]: This is outside the journal's usage contract rather than a defect. `prune` returns `Err` when removal fails, and a failed mutable storage operation is fatal for the instance (workspace convention): the caller must not keep using it, so the subsequent `sync` in this scenario is already a contract violation.

Even for a caller that continues anyway, no state in this scenario loses acknowledged data, and recovery reconciles all of them. If the stale boundary is re-persisted while the survivors are intact, the prune un-happens — legal, because the errored prune was never acknowledged. If removal had progressed past the re-persisted boundary's blob, init sees the gap and adopts the physical boundary, completing the prune. A boundary ahead of physical blobs is precisely the interrupted-prune state this PR's recovery is designed to finish (and the in-memory inconsistency after a failed removal predates this change; the up-front persist is what makes these states recoverable at all).

File: storage/src/journal/contiguous/variable.rs [hunk 2]
```diff
@@ -1880,12 +1920,22 @@ impl<E: Context, V: CodecShared> Journal<E, V> {
...
+                )));
+            }
+
+            // Otherwise this is a fresh partition or a journal with no retained durable
+            // items. Clear (rather than prune) the offsets so bounds collapse even when the
+            // size is mid-blob.
             let size = offsets_bounds.end;
             if !offsets_bounds.is_empty() {
-                warn!("crash repair: clearing offsets to {size} (prune-all crash)");
+                warn!("crash repair: clearing offsets to {size} (no data blobs)");
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/variable.rs:1902]: Good catch — this was a real issue. Fixed in 738594da4 by applying the recovery-watermark corruption check to every itemless-data state before distinguishing no files from a surviving empty tail. The new test_variable_recovery_rejects_item_blob_loss_with_surviving_tail regression test covers the reported case and verifies that corruption remains stable across retried inits.

File: storage/src/journal/contiguous/fixed.rs [hunk 3]
```diff
@@ -1053,6 +1062,13 @@ impl<E: Context, A: CodecFixedShared> Journal<E, A> {
...
         self.blobs.prune(min_blob).await?;
```
Comment (danlaine) [inline storage/src/journal/contiguous/fixed.rs]: Other cancel-safety related work is ongoing, so this code may change again, but I think this finding is valid:

```
Cancelled prune + continued use can arm a stable corruption verdict (new in this PR)

If a prune future is dropped mid-removal and the journal is then used normally, the next
reopen can report unrecoverable corruption — from legal API use only. The chain:

1. prune persists the new boundary first, but updates self.bounds.start only after
blobs.prune returns. A dropped future leaves the in-memory boundary stale while the
durable one is already ahead.
2. Removals are now concurrent, so a dropped future can leave a gap below the committed
boundary (a newer blob unlinked, an older one not) — the old oldest-first loop could only
leave a contiguous prefix removed.
3. A later sync() persists the stale bounds.start, lowering the durable boundary back
below the gap and reclassifying the condemned blobs as retained.
4. On the next reopen, recovery correctly rejects that shape — either the gap sits in the
acknowledged region (watermark exceeds the recovered size) or the boundary blob itself is
missing (aligned boundary can never lag). Both verdicts are right for the disk shape; the
problem is the process can now self-inflict it.

Pre-PR, every ingredient was absent: removals couldn't gap, ordinary prune/sync never
persisted an aligned boundary, and recovery derived the boundary from the oldest survivor —
so the same cancel-then-continue usage recovered cl

The committed boundary only means "everything below was condemned" if nothing can lower it
while removals are outstanding. Suggested fix: assign bounds.start = new_boundary (and
clamp dirty_from_blob) immediately after persist suobs —
matching the state-before-I/O ordering Writable::prune already uses internally. A cancelled
prune then leaves memory agreeing with the durable  files below it
are exactly the interrupted-prune shape recovery already completes. The variable journal
needs the same treatment via its offsets journal (ournal prune; a
later variable.sync() → offsets.sync() persists the stale offsets boundary the same way).
```

```suggestion
        self.bounds.start = new_boundary;
        self.blobs.prune(min_blob).await?;
```

Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs]: Confirmed and fixed — the finding is valid and exactly as you traced it: with removals both concurrent and gated behind an up-front boundary commit, a dropped prune future left memory stale against the durable checkpoint, and a later `sync()` from the still-usable journal re-persisted the stale boundary beneath a possible removal gap — a shape recovery then correctly rejects.

Took your suggested ordering (21f5497d4, refined in c5a7faba4): both journals adopt the committed boundary in memory at the same transition that persists it, before any removal starts. In the fixed journal that's immediately after the checkpoint persist (plus the `dirty_from_blob` clamp). The variable journal's bounds update is coupled to the offsets journal's commit transition via an internal `on_commit` callback that fires synchronously between the persist and the removals — a future dropped anywhere inside the offsets prune cannot leave the outer journal reporting bounds whose offsets are already pruned. A drop during the persist itself is benign: nothing has been removed yet, so a later sync re-persisting the old boundary just un-does an unacknowledged prune, gap-free.

Regressions in both journals pin the dropped-at-removal state via test-only hooks that run the production commit stage and stop at the removal await — no error is surfaced, matching cancellation semantics, so the tests' continued use is legal. The fixed test manufactures the one-unlinked-blob gap, then syncs and reopens to the completed prune; the variable test also pins the read contract: after the simulated drop, retained reads succeed and positions below the boundary report `ItemPruned` rather than sitting inside `bounds()`.

**Review by danlaine — DISMISSED** [review]

**Review by danlaine — DISMISSED** [review]

File: storage/src/journal/contiguous/fixed.rs [hunk 4]
```diff
@@ -582,55 +586,76 @@ impl<E: Context, A: CodecFixedShared> Inner<E, A> {
...
+            .collect();
+        if !stale.is_empty() {
+            warn!(
+                count = stale.len(),
+                boundary, "crash repair: completing interrupted prune"
+            );
+            for blob in &stale {
+                drop(pending.remove(blob));
             }
+            try_join_all(stale.into_iter().map(|blob| partition.remove(blob))).await?;
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs:717]: Fixed in c08110963: the classification now fires when any blob survives (pending.first_key_value(), not only a survivor at/after the boundary), so the prefix-only shape gets a stable corruption verdict before any file is unlinked. Added `test_fixed_recovery_missing_boundary_with_only_stale_prefix`, which fails against the previous code by observing the deleted prefix.

File: storage/src/journal/contiguous/variable.rs [hunk 5]
```diff
@@ -3092,15 +3142,10 @@ mod tests {
...
 
-    /// Test that init aligns state when data is pruned/lost but offsets survives.
-    ///
-    /// This handles both:
-    /// 1. Crash during prune-all (data pruned, offsets not yet)
-    /// 2. External data partition loss
-    ///
-    /// In both cases, we align by pruning offsets to match.
+    /// Losing the data partition while the watermark proves durable retained items is
+    /// corruption, stable across retries -- not a completed prune-all to accept silently.
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/variable.rs]: Reworded in c08110963 (plus one other `--` in fixed.rs).

File: storage/src/journal/contiguous/fixed.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs:1142]: Resolved by the current design: `commit_prune` adopts `bounds.start` before `checkpoint.persist`, and both happen before any blob removal, so a later sync can never regress the committed boundary.

File: storage/src/journal/contiguous/variable.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/variable.rs:1902]: Resolved: `align_empty` now runs the watermark-vs-boundary corruption check as its first statement, before any branch or repair, so the shape with surviving empty blob files is covered.

File: storage/src/journal/contiguous/fixed.rs [hunk 6]
```diff
@@ -506,6 +513,16 @@ impl<E: Context, A: CodecFixedShared> Inner<E, A> {
...
 
+        // Complete any interrupted prune (a committed boundary ahead of physical storage) and
+        // reconcile the pruning boundary before walking blob lengths.
+        let pruning_boundary = Self::reconcile_pruning_boundary(
+            &partition,
+            &mut pending,
+            cfg.items_per_blob.get(),
+            checkpoint.boundary_hint(),
+        )
+        .await?;
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs:530]: On reflection this is out of the crash model rather than a defect, so we're intentionally not hardening against it. A blob below a committed pruning boundary was synced before the prune, so it cannot be crash-torn; reaching this state requires external corruption of a condemned blob that also happens to be one of the two newest.

Recovery handles the crash model exactly (the durable boundary completes the interrupted prune) and otherwise fails loud, uniformly, on anything below the watermark it cannot explain, instead of fine-tuning which corrupted-and-condemned shapes to tolerate. Reporting `Corruption` here is that deliberate loud failure on external damage.

We prototyped reordering the scan to run after prune reconciliation (so condemned blobs are removed first and never inspected), but that adds recovery complexity purely to swallow external corruption of data that is about to be deleted, so we backed it out.

File: storage/src/journal/contiguous/fixed.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs:717]: `reconcile_pruning_boundary` already classifies a missing aligned boundary blob as `Corruption` before removing the stale prefix, via the `boundary.is_multiple_of(items_per_blob) && !pending.contains_key(&boundary_blob) && first_key_value().is_some()` guard (comment: "Classify before removing the stale prefix below the boundary, so a failed init leaves every surviving blob untouched"). So no destructive action precedes the corruption verdict for the reported case.

This shape is also out of the crash model: an interrupted prune never deletes the boundary blob, so a missing one is external loss, correctly reported loud. The only path that unlinks before a later verdict is a mid-blob boundary, which by design adopts the oldest surviving blob (a legacy `init_at_size` boundary can lag a completed prune) and removes only condemned below-boundary data. No acknowledged data is lost and the verdict is stable across retried inits.

File: storage/src/journal/contiguous/fixed.rs [hunk 7]
```diff
@@ -600,21 +617,17 @@ impl<E: Context, A: CodecFixedShared> Inner<E, A> {
...
+    /// The pruning boundary has already been reconciled with physical blob state: an ahead
+    /// boundary completed an interrupted prune, and a boundary whose blob is missing while later
+    /// blobs survive was rejected as corruption. A watermark beyond the recovered size is
+    /// corruption. The caller persists the checkpoint before applying the returned repair (see
+    /// comment at the call site).
     fn recover_bounds(
         pending: &BTreeMap<u64, Writer<E::Blob>>,
         items_per_blob: u64,
-        boundary_hint: Option<u64>,
+        pruning_boundary: u64,
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply storage/src/journal/contiguous/fixed.rs:628]: Good observation that legacy recovery can preserve and re-persist `boundary > watermark`; I clarified the invariant and named that exception in e076f47ef.

This is intentional under the PR compatibility model, rather than a current-prune safety hole. The stale pair can only be inherited from pre-#4235 state. The supported upgrade procedure requires a clean `sync`/drain first, which records the journal end as the watermark, so `watermark >= boundary`. Every prune produced by this PR likewise syncs the data and atomically persists `(boundary, bounds.end)` before unlinking anything.

Using `max(watermark, boundary)` for the fixed hole check would not prove the retained suffix durable; it would only raise the floor to the boundary and could overstate what a legacy checkpoint acknowledged. Fully supporting an unsynced legacy prune would require another durable migration marker to distinguish it from an ordinary unacknowledged suffix. Under the documented upgrade assumption, the under-protected legacy state is unreachable, so no recovery-code change is needed.

**Comment by roberto-bayardo** [comment]
@patrick-ogrady does your recent PR supercede this one?  Looks like it does something similar.

**Comment by patrick-ogrady** [comment]
> @patrick-ogrady does your recent PR supercede this one?  Looks like it does something similar.

Something similar but in the archive layer.
